### ISSUE TITLE
DX: Improve serverless plugins discoverability & install process

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#2861]
# This is a Feature Proposal

## Description

Add a `serverless plugins` command so that Serverless plugins can be searched and installed (wrapper around `npm`) through the CLI. This is a big DX gain as the user doesn't need to leave the CLI to search and install plugins (it feels less nodeish).

@DavidWells and I also talked about way to offer a website where one can search through all the Serverless plugins and services. He's also working on an integration in the dashboard (so the `plugins.json` file (see below) is the single point of truth for all Serverless plugins).

**search**

The `--search` option simply looks in the registry here https://github.com/serverless/community-plugins/blob/master/plugins.json and searches through the plugins.

**install**

The `--install` option will call `npm install --save` with the name of the plugin and adds the plugin to the plugin section of the services `serverless.yml` file. A message will be shown that the user should check if the order of the plugin addition is correct (the newly installed plugin will be pushed to the end of the `plugins` array).

**remove**

The `--remove` option will call `npm uninstall --save` and removes the plugin from the services `serverless.yml` file.

---

This new `plugins` command might be reused by #2821 / #1260.

**Comment by nikgraf** [comment]
I think it's important it install plugins only as `--save-dev`.

**Comment by mthenw** [comment]
I think that `search` and `install` should be subcommands not options. And also command should be called `plugin`.

```
serverless plugin install
serverless plugin search
```

**Comment by dougmoscrop** [comment]
How does this intersect with[ plugin tree visualization?](https://github.com/serverless/serverless/pull/2821) `serverless plugin [list|visualize|tree]` in terms of plural vs. not plural?

- search returns multiple plugins
- install could very well install multiple at one time (`serverless plugins install -n foo -n bar` which is how npm works)

**Comment by DavidWells** [comment]
I like the singular version

```
serverless plugin
```

- `serverless plugin install` - downloads plugin and automagically adds to `serverless.yml`. (note a yaml AST parser is needed for this to keep peoples comments in place if they have them
- `serverless plugin list` - list all available/known plugins from the [registry](https://github.com/serverless/community-plugins/blob/master/plugins.json)
- `serverless plugin search` - search for match out of list. If no match found, suggest categories/tags that are going to get added to the [plugin meta](https://github.com/serverless/community-plugins/blob/master/plugins.json)
- `serverless plugin visualize` - I freakin love @dougmoscrop's setup here. Very handy for understanding how the plugin system works.

**Comment by eahefnawy** [comment]
Will be moved to an external plugin (not in core)

**Comment by pmuens** [comment]
Closing this since we'll provide another way to solve this problem soon.

**Comment by DavidWells** [comment]
Why are we closing this?

**Comment by pmuens** [comment]
Closing this since it was implemented via #4046
